Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -6645,10 +6647,19 @@ static bool frameSandboxAllowsOpeningExternalCustomProtocols(SandboxFlags sandbo
 }
 #endif
 
-void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& process, PageIdentifier webPageID, WebFrameProxy& frame, NavigationActionPolicyParameters&& parameters, CompletionHandler<void(PolicyDecision&&)>&& completionHandler)
+void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& process, PageIdentifier pageID, WebFrameProxy& frame, NavigationActionPolicyParameters&& parameters, CompletionHandler<void(PolicyDecision&&)>&& completionHandler)
+{
+    CoroutineCaller* caller = new CoroutineCaller();
+    caller->setCoroutine([this, protectedThis = Ref { *this }, caller, process = WTFMove(process), pageID, frame = Ref { frame }, parameters = WTFMove(parameters), completionHandler = WTFMove(completionHandler)] () mutable -> Task {
+        completionHandler(co_await awaitableDecidePolicyForNavigationAction(WTFMove(process), pageID, frame, WTFMove(parameters)));
+        delete caller;
+    });
```

Comment:
This pattern feels a little awkward... I think I see why `CoroutineCaller` needs to stay alive, but maybe `CoroutineCaller` should know to clean up after itself when the `completionHandler` eventually gets called (after `awaitableDecidePolicyForNavigationAction` returns...)?

Reply:
Yeah, I agree this needs to have a better abstraction that basically does this same thing.